File: mwc/romana/relic/b/lib/libc/string/i8086/strspn.m
//////////
/ i8086 C string library.
/ strspn()
/ ANSI 4.11.5.6.
//////////

//////////
/ size_t
/ strspn(String1, String2)
/ char *String1, *String2;
/
/ Return the length of the initial segment of String1 which consists of
/ characters in String2
//////////

#include <larges.h>

String1	=	LEFTARG
String2	=	String1+DPL

	Enter(strspn_)
	Les	di, String1(bp)	/ String1 address String1 ES:DI
	mov	dx, di		/ Save for length calc
	cld
1:	movb	ah, Pes (di)	/ Get char from String 1
	orb	ah, ah
	je	3f		/ If equal done

	Lds	si, String2(bp)	/ address of String2 DS:SI
2:	lodsb			/ From character String1 AL
	orb	al, al
	je	3f		/ End of scan no match means done
	cmpb	ah, al
	jne	2b		/ Try next in String2
	inc	di		/ Try next in String1
	jmp	1b
3:	mov	ax, di
	sub	ax, dx
	Leave

File: mwc/romana/relic/d/lib/libc/string/strspn.c
/*
 * strspn.c
 * ANSI 4.11.5.6.
 * Search string for string characters.
 */

#include <string.h>

size_t *strspn(s1, s2) char *s1; char *s2;
{
	register char *cp1, *cp2;
	register char c1, c2;

	for (cp1 = s1; c1 = *cp1; cp1++) {
		for (cp2 = s2; c2 = *cp2; cp2++)
			if (c1 == c2)
				break;
		if (c2 == '\0')
			break;
	}
	return((size_t)(cp1 - s1));
}
